src/foam/nanos/crunch/JavaCrunchService.java

```diff
@@ -0,0 +1,140 @@
+        grantPath.remove(previousIndex);
+
+        // Remove previously stored index of capability
+        alreadyListed.remove(sourceId);
+
+        // Shift remembered indexes, now that grantList has been shifted
+        Map<String,Integer> newAlreadyListed = new HashMap<String,Integer>();
+        for ( Map.Entry<String,Integer> entry : alreadyListed.entrySet() ) {
+          int newIndex = entry.getValue();
+          if ( newIndex > previousIndex ) newIndex--;
+        }
+        alreadyListed = newAlreadyListed;
+      }
+
+      // Add capability to grant path, and remember index in case it's replaced
+      Capability cap = (Capability) capabilityDAO.find(sourceId);
+
+      alreadyListed.put(sourceId, grantPath.size());
+      grantPath.add(cap);
+
+      // Enqueue prerequisites for adding to grant path
+      List prereqs = ( (ArraySink) prerequisiteDAO
+        .where(EQ(CapabilityCapabilityJunction.SOURCE_ID, sourceId))
+        .select(new ArraySink()) ).getArray();
+      for ( Object prereqObj : prereqs ) {
+        CapabilityCapabilityJunction prereq =
+          (CapabilityCapabilityJunction) prereqObj;
+        nextSources.add(prereq.getTargetId());
+      }
+    }
+
+    Collections.reverse(grantPath);
+    return grantPath;
+  }
+
+  public UserCapabilityJunction getJunction(X x, String capabilityId) {
+    User user = ((Subject) x.get("subject")).getUser();
+    User realUser = ((Subject) x.get("subject")).getRealUser();
+
+    Predicate acjPredicate = INSTANCE_OF(AgentCapabilityJunction.class);
+    Predicate targetPredicate = EQ(UserCapabilityJunction.TARGET_ID, capabilityId);
+    try {
+      DAO userCapabilityJunctionDAO = (DAO) x.get("userCapabilityJunctionDAO");
+
+      if ( user != realUser ) {
+        // Check if a ucj implies the subject.user has this permission
+        Predicate userPredicate = AND(
+          NOT(acjPredicate),
+          EQ(UserCapabilityJunction.SOURCE_ID, user.getId())
+        );
+        UserCapabilityJunction ucj = (UserCapabilityJunction)
+          userCapabilityJunctionDAO.find(AND(userPredicate,targetPredicate));
+        if ( ucj != null ) {
+          return ucj;
+        }
+      }
+
+      // Check if a ucj implies the subject.realUser has this permission
+      {
+        Predicate userPredicate = AND(
+          NOT(acjPredicate),
+          EQ(UserCapabilityJunction.SOURCE_ID, realUser.getId())
+        );
+        UserCapabilityJunction ucj = (UserCapabilityJunction)
+          userCapabilityJunctionDAO.find(AND(userPredicate,targetPredicate));
+        if ( ucj != null ) {
+          return ucj;
+        }
+      }
+
+      // Check if a ucj implies the subject.realUser has this permission in relation to the user
+      {
+        Predicate userPredicate = AND(
+          acjPredicate,
+          EQ(UserCapabilityJunction.SOURCE_ID, realUser.getId()),
+          EQ(AgentCapabilityJunction.EFFECTIVE_USER, user.getId())
+        );
```

Comment:
Your conditions could build a predicate and then you could call the ucj DAO once at the end. You could also build out the predicate through ternary expressions if you really wanna shorten it.

Reply:
Sick! Definitely doing that, thanks!